Introduce a JettyClientEngine to use jetty-client with resteasy JAX-RS Client

stevenschlansker opened the pull request:
Jetty has a mature, efficient, and very nice client that is very handy if you are already using Jetty as your server and would prefer a single stack solution over introducing AsyncHttpClient.

stevenschlansker commented:
Note that I've added a little cleanup around the executor handling -- right now we require `ExecutorService` but never really use it.  In Jetty's case we already have an Executor we can reuse (but unfortunately it is *not* an ExecutorService).  This is not strictly necessary for this feature but it led to some nice cleanup that could opportunistically make 4.0 (and otherwise probably can't come until 5)

pennello commented on resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/engines/JettyClientEngine.java:
```diff
@@ -0,0 +1,223 @@
+            try {
+                client.start();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
```
Is this how this library prefers to handle the construction of these engines?  Seems unfortunate to depend on using an object that throw a checked exception, but then not have handling for checked exceptions for the construction of these engines.

pennello commented on resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/engines/JettyClientEngine.java:
```diff
@@ -0,0 +1,223 @@
+
+    @Override
+    public HostnameVerifier getHostnameVerifier() {
+        throw new UnsupportedOperationException();
```
Does the lack of an implementation of this method merit documentation?

stevenschlansker replied:
Not sure what the consequence of this is.  I mirrored the Apache implementation.

pennello commented on resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/engines/JettyClientEngine.java:
```diff
@@ -0,0 +1,223 @@
+        try {
+            return future.get(client.getIdleTimeout(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            future.cancel(true);
```
And reset the interrupt flag?

pennello commented on resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/engines/JettyClientEngine.java:
```diff
@@ -0,0 +1,223 @@
+            future.cancel(true);
+            throw clientException(e, null);
+        } catch (ExecutionException | TimeoutException e) {
+            throw clientException(e.getCause(), null); // NOPMD
```
What was PMD complaining about here?

pennello commented on resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/engines/JettyClientEngine.java:
```diff
@@ -0,0 +1,223 @@
+     * Implementation note: due to lack of asynchronous message decoders the request must either be buffered,
+     * or it must have a {@code null} extractor and type parameter {@code <T>} must be {@link ClientResponse},
+     * which will read the data through its stream.  It is not possible to use the synchronous JAX-RS message
+     * decoding infrastructure without buffering or spinning up auxiliary threads (arguably more expensive than buffering).
```
The argument would depend on the size of the payload.

stevenschlansker replied:
Yeah, I agree -- we may end up adding the threaded option later, but it's not needed to ship the first version.

pennello commented on resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/engines/JettyClientEngine.java:
```diff
@@ -0,0 +1,223 @@
+                cr.setHeaders(extract(response.getHeaders()));
+                if (!buffered) {
+                    if (extractor != null) {
+                        throw new IllegalStateException("TODO: ResultExtractor is synchronous and may not be used without buffering.");
```
This is neat.

pennello commented on resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/engines/JettyClientEngine.java:
```diff
@@ -0,0 +1,223 @@
+
+            @Override
+            public void onSuccess(Response response) {
+                if (buffered) {
```
In the case that we're unbuffered, we'll have a `null` extractor, so is that why it's okay to do nothing here in that case?

stevenschlansker replied:
Yup, exactly.

stevenschlansker commented:
Thanks for the feedback.  I'm working on the test failures and will have a fix for that and the feedback soon.

stevenschlansker commented on testsuite/arquillian-utils/pom.xml:
```diff
@@ -191,7 +191,7 @@
         <dependency>
             <groupId>org.slf4j</groupId>
-            <artifactId>slf4j-nop</artifactId>
+            <artifactId>slf4j-simple</artifactId>
```
I'm not sure why the test utils are pulling in `slf4j-nop` -- that just leads to no log output, making debugging difficult.

stevenschlansker commented:
Hi @ronsigal @asoldano sorry for the direct ping, but I am wondering -- is this a contribution RESTEasy could accept?  I'm excited to get it into 4.0 as early as possible!

asoldano commented:
@stevenschlansker np at all for the direct ping. Yes, this is a contribution we can accept, thanks. It probably just needs to be cleaned up a little bit and enriched with some documentation in the docbook, but in any case it looks interesting :-)
While we're here, this would be your first contribution since I'm leading the project; we started requiring that all contributions be made under the terms of the ASL2.0 License: http://www.apache.org/licenses/LICENSE-2.0 . Do you agree to these terms?

stevenschlansker commented:
Yes, the code is original and I release it as Apache2.

stevenschlansker commented:
hi @asoldano we've been running this in production for a while now and haven't found additional issues, I introduced a brief documentation blurb, and rebased it against 4.0.0.Beta2

What else could I do to move this forward?  Thanks!

asoldano commented:
@stevenschlansker thanks, I've created https://issues.jboss.org/browse/RESTEASY-1862 , will review this again and eventually merge in the next weeks.

asoldano commented:
@stevenschlansker , I'm closing this and replacing it with https://github.com/resteasy/Resteasy/pull/1495 , which is a rebased version of this PR (to solve the conflict); it has the same commits as this PR, plus a minor fix to the documentation. If the testsuite passes on #1495, I'll merge.
